controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -300,49 +332,53 @@ func (r *KubeadmControlPlaneReconciler) initializeControlPlane(ctx context.Conte
 		ownerRef,
 	)
 	if err != nil {
-		return errors.Wrap(err, "Failed to generate bootstrap config")
+		errs = append(errs, errors.Wrap(err, "Failed to generate bootstrap config"))
 	}
 
-	// Create the Machine
-	err = r.MachineGenerator.GenerateMachine(
-		ctx,
-		r.Client,
-		kcp.Namespace,
-		kcp.Name,
-		cluster.Name,
-		kcp.Spec.Version,
-		infraRef,
-		bootstrapRef,
-		generateKubeadmControlPlaneLabels(cluster.Name),
-		ownerRef,
-	)
-	if err != nil {
-		logger.Error(err, "Unable to create Machine")
-		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "FailedCreateMachine", "Failed to create machine: %v", err)
-
-		errs := []error{err}
+	// Only proceed to generating the Machine if we haven't encountered an error
+	if len(errs) == 0 {
+		err := r.MachineGenerator.GenerateMachine(
+			ctx,
+			r.Client,
+			kcp.Namespace,
+			kcp.Name,
+			cluster.Name,
+			kcp.Spec.Version,
+			infraRef,
+			bootstrapRef,
+			generateKubeadmControlPlaneLabels(cluster.Name),
+			ownerRef,
+		)
+		if err != nil {
+			errs = append(errs, errors.Wrap(err, "Failed to create Machine"))
+		}
+	}
 
+	// If we encountered any errors, attempt to clean up any dangling resources
+	if len(errs) > 0 {
+		// Since we return early if there is an error with cloning the infra config,
```

Comment:
We can move this cleanup logic into a different private method, it should be easier to read that way

Reply:
Done, I'm not quite sure how the errors will look with potentially aggregating an already aggregated error.